cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i

```diff
@@ -286,32 +303,136 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute
         c                            = 2 * lipschitz + displacement;
     }
 
+    const size_t nRowsInBlock = 512;
+    const size_t nDataBlocks  = n / nRowsInBlock + !!(n % nRowsInBlock);
+
+    TArrayScalable<algorithmFPType, cpu> values;
+    if (valueNT)
+    {
+        values.reset(nDataBlocks);
+        DAAL_CHECK_MALLOC(values.get());
+    }
+
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, nRowsInBlock, nClasses);
+    TlsMem<algorithmFPType, cpu> tlsLogP(nRowsInBlock * nClasses);
+    TArrayScalable<algorithmFPType, cpu> grads;
+    if (gradientNT)
+    {
+        DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, nDataBlocks, nBeta);
+        grads.reset(nDataBlocks * nBeta);
+        DAAL_CHECK_MALLOC(grads.get());
+    }
+    TlsMem<algorithmFPType, cpu> tlsSoftmaxSum(nClasses);
+    const algorithmFPType div = static_cast<algorithmFPType>(1) / static_cast<algorithmFPType>(n);
+    const bool bL1            = parameter->penaltyL1 > 0;
+    const bool bL2            = parameter->penaltyL2 > 0;
+    const bool interceptFlag  = parameter->interceptFlag;
+
     if (valueNT || gradientNT || hessianNT)
     {
-        //f = X*b + b0
-        applyBetaThreaded(x, b, f.get(), n, nClasses, p, parameter->interceptFlag);
+        daal::threader_for(nDataBlocks, nDataBlocks, [&](size_t iBlock) {
+            const size_t iStartRow      = iBlock * nRowsInBlock;
+            const size_t nRowsToProcess = (iBlock == nDataBlocks - 1) ? n - iBlock * nRowsInBlock : nRowsInBlock;
+
+            const algorithmFPType * const xLocal = x + iStartRow * p;
+            const algorithmFPType * const yLocal = y + iStartRow;
+
+            algorithmFPType * const fPtrLocal = f.get() + iStartRow * nClasses;
 
-        //f = softmax(f)
-        softmaxThreaded(f.get(), f.get(), n, nClasses);
+            //f = X*b + b0
+            applyBeta(xLocal, b, fPtrLocal, nRowsToProcess, nClasses, p, interceptFlag);
 
-        const bool bL1 = (parameter->penaltyL1 > 0);
-        const bool bL2 = (parameter->penaltyL2 > 0);
+            //f = softmax(f)
+            algorithmFPType * softmaxSums = nullptr;
+
+            if (interceptFlag && gradientNT)
+            {
+                softmaxSums = tlsSoftmaxSum.local();
+                softmax(fPtrLocal, fPtrLocal, nRowsToProcess, nClasses, softmaxSums, yLocal);
+            }
+            else
+            {
+                softmax(fPtrLocal, fPtrLocal, nRowsToProcess, nClasses, nullptr, nullptr);
+            }
+            const algorithmFPType * const fixedSoftmaxSums = softmaxSums;
+
+            if (valueNT)
+            {
+                algorithmFPType * const logP = tlsLogP.local();
+                daal::internal::Math<algorithmFPType, cpu>::vLog(nRowsToProcess * nClasses, fPtrLocal, logP);
```

Comment:
check overflow

Reply:
nRowsToProcess * nClasses overflow is checked implicitly. https://github.com/oneapi-src/oneDAL/blob/ce6d3d851830f225d888eeca5a9f71f0538c5f1d/cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i#L316 checks nRowsInBlock * nClasses overflow